File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def split_record(record):

    res = record.split()
    if len(res) == 3:
        return res[0], res[1], res[2]
    else:
        return res[0], res[1], None


def solution(record):
    answer = []
    log = {}

    for i in record:
        action, uid, name = split_record(i)
        if action == "Enter":
            log[uid] = name
            answer.append(f"{uid}님이 들어왔습니다.")
        elif action == "Leave":
            answer.append(f"{uid}님이 나갔습니다.")
        elif action == "Change":
            log[uid] = name

    for i in range(len(answer)):
        uid = answer[i].split("님")[0]
        answer[i] = answer[i].replace(uid, log[uid])
    return answer

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        String[] answer = {};
        Map<String,String> msg = Map.of(
                "Enter","님이 들어왔습니다.",
                "Leave", "님이 나갔습니다."
        );

        Map<String,String> map = new HashMap<>();
        for(String cmd : record) {
            String[] tmp = cmd.split(" ");
            if(tmp.length == 3){ // Enter or Change
                map.put(tmp[1],tmp[2]); // uid : name
            }
        }

        ArrayList<String> ans = new ArrayList<>();
        for(String cmd : record){
            String[] log = cmd.split(" ");
            if(log[0].equals("Enter") || log[0].equals("Leave")){
                String tmp = map.get(log[1]).concat(msg.get(log[0]));
                ans.add(tmp);

            }
        }
        return ans.toArray(new String[0]);
    }
}